drivers/bluetooth/hci/h4.c

```diff
@@ -22,9 +22,19 @@
 #include <zephyr/bluetooth/hci.h>
 #include <zephyr/drivers/bluetooth/hci_driver.h>
 
-#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HCI_DRIVER)
-#define LOG_MODULE_NAME bt_driver
-#include "common/log.h"
+#include <zephyr/logging/log.h>
+
+#ifdef CONFIG_BT_DEBUG_LOG
+#ifdef CONFIG_BT_DEBUG_HCI_DRIVER
+#define LOG_LEVEL LOG_LEVEL_DBG
+#else
+#define LOG_LEVEL LOG_LEVEL_INF
+#endif
+#else
+#define LOG_LEVEL LOG_LEVEL_NONE
+#endif
+
+LOG_MODULE_REGISTER(bt_driver, LOG_LEVEL);
```

Comment:
Since you have to repeat these 13 lines in every single c-file, maybe it would still make sense to keep a `common/log.h` header with a helper for this, e.g.:

```c
#include "common/log.h"
BT_LOG_REGISTER(bt_driver, IS_ENABLED(CONFIG_BT_DEBUG_HCI_DRIVER));
```

The `BT_LOG_REGISTER()` API in `common/log.h` could be defined something like (completely untested):
```c
#ifdef CONFIG_BT_DEBUG_LOG
#define BT_LOG_LEVEL_1 LOG_LEVEL_DBG
#define BT_LOG_LEVEL_0 LOG_LEVEL_INF
#else
#define BT_LOG_LEVEL_0 LOG_LEVEL_NONE
#endif

#define BT_LOG_REGISTER(name, enable) \
	LOG_MODULE_REGISTER(name, _CONCAT(BT_LOG_LEVEL_, enable))
```

Reply:
Generally in BT we've had something similar to 
```c
#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_SMP)
#define LOG_MODULE_NAME bt_smp
#include "common/log.h"
```

Which I guess could be transformed into something a bit shorter like
```c
#include <zephyr/logging/log.h>

LOG_MODULE_REGISTER(bt_smp, IS_ENABLED(CONFIG_BT_DEBUG_SMP) ? LOG_LEVEL_DBG : CONFIG_BT_LOG_LEVEL);
```

The alternative is to actually define a log level for each file/module instead.

Reply:
@Thalley are you sure `LOG_MODULE_REGISTER()` accepts a conditional statement as its second argument? I assumed it needs to be a directly valid preprocessor symbol, which is what I based my proposal on.

Reply:
Zephyr has a preprocessor ternary: `COND_CODE_1`.

Reply:
It would appear not. It doesn't fail at building, but it does fail at runtime. 

In that case, your suggestion is possibly better, but perhaps mine could be modified to

```c
#include <zephyr/logging/log.h>
#if defined(CONFIG_BT_DEBUG_SMP)
#define SMP_LOG_LEVEL LOG_LEVEL_DBG
#else 
#define SMP_LOG_LEVEL CONFIG_BT_LOG_LEVEL
#endif

LOG_MODULE_REGISTER(bt_smp, SMP_LOG_LEVEL);
```

Which does work, but isn't much better than what @theob-pro is suggesting

Reply:
You can write it like the following:

```c
LOG_MODULE_REGISTER(bt_smp, COND_CODE_1(CONFIG_BT_DEBUG_SMP, (LOG_LEVEL_DBG), (CONFIG_BT_LOG_LEVEL)));
```

Note that @theob-pro's code is a true refactor. Our suggestion here changes the meaning of the kconfigs slightly. (But I don't think that's necessarily an issue.)

Reply:
> Note that @theob-pro's code is a true refactor. Our suggestion here changes the meaning of the kconfigs slightly. (But I don't think that's necessarily an issue.)

Is it though? 

The current code uses 

```c
#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HCI_DRIVER)
#define LOG_MODULE_NAME bt_driver
#include "common/log.h"
```

and in log.h we have

```c
#if BT_DBG_ENABLED
#define LOG_LEVEL LOG_LEVEL_DBG
#else
#define LOG_LEVEL CONFIG_BT_LOG_LEVEL
#endif
```

Wouldn't that indicate that `CONFIG_BT_DEBUG_HCI_DRIVER=y` => `LOG_LEVEL_DBG` and `CONFIG_BT_DEBUG_HCI_DRIVER=n` => `CONFIG_BT_LOG_LEVEL` which is also what I am suggesting. 

The code suggested by @theob-pro 
```c

#ifdef CONFIG_BT_DEBUG_LOG
#ifdef CONFIG_BT_DEBUG_HCI_DRIVER
#define LOG_LEVEL LOG_LEVEL_DBG
#else
#define LOG_LEVEL LOG_LEVEL_INF
#endif
#else
#define LOG_LEVEL LOG_LEVEL_NONE
#endif
```

Sets `LOG_LEVEL` to `LOG_LEVEL_INF` if `CONFIG_BT_DEBUG_HCI_DRIVER=n` which is different that the current code (it should be `CONFIG_BT_LOG_LEVEL`). 

Or am I missing something?

Reply:
Ah. I didn't see the definition of `config BT_LOG_LEVEL` because it's instantiated from a template.

I see now that you are correct. @Thalley's proposal is the correct refactoring. Agree, @theob-pro?